doc/design/python_api.md

```diff
@@ -0,0 +1,156 @@
+    def __init__(self, shape, dtype="float32", name=None, block=None):
+        if name is None:
+            name = unique_name_generator()
+        self.name = name
+        self.block = block
+        # build C++ Protobuf object
+        self.cpp_var_desc_ptr = ...
+        self.op = None
+
+    def shape(self):
+        cpp_shape = self.cpp_var_desc_ptr.shape()
+        return [None if elem < 0 else elem for elem in cpp_shape]
+```
+
+The Protobuf object should be created in C++ not Python because it is needed by infershape, and infershape is implemented by C++ code. The C++ Protobuf object is accessible for Python through the `cpp_var_desc_ptr`, just like how `shape()` function does.
+
+The user is allowed to build a variable without specifying its name. If so, it is going to be assigned with an automatically generated unique name.
+
+### Parameter
+
+The parameter is a kind of special variable. They need to be initialized at the very beginning and updated after each batch training. So if a variable is a parameter, our compiler will add an initializer op and an optimizer op for it during the building process of computation graph. Apart from these, there is no more difference between variable and parameter. In other words, 'parameter' is only a label attached to variables, to tell the compiler these ones require additional processing.
+
+```python
+class Parameter(Variable):
+    def __init__(self, trainable, initialize_attrs, optimize_attrs):
+        pass
+```
+
+The class `Parameter` is derived from class `Variable`. In addition to variables have, parameters are able to hold their initializing and updating information. A parameter's `self.op` will always be `None` because it can never be an operator's output.
+
+
+## Layer Functions
+
+A layer is a Python function. When it is invoked, it creates a series of operators and variables then inserts them into the block. It is something like the macro in C++. It is called 'Layer' because the combination of added operators acts just like what a neural network layer does. 
+
+Here are examples of how to write a data layer and FC layer:
+
+### Data Layer
+
+```python
+def data_layer(name, type, block=None):
+    if block is None:
+        block = g_block
+    # type = dense_vector(size=10) / integer_value(range=10)
+    return block.create_global_var(
+            name=name, 
+            shape=[None] + type.dims(), 
+            dtype=type.dtype)
+
+``` 
+
+Before building new variables, we need to specify which block to use. If we don't, the default one `g_block` will be used. In the above `data_layer` code, a variable is created and be inserted into the root block to make it global. This variable is going to be used as input data of the whole network.
+
+### FC Layer
+
+```python
+def fc_layer(input, size, block=None, ...):
+    if block is None:
+        block = g_block
+    w = block.create_parameter(...)
+    b = block.create_parameter(...)
+    out = stack.create_var()
```

Comment:
what is `stack` for?

Reply:
Actually, this is a topy, should be `block`